File: SoinnPlus.m
% SOINN+ - Self-Organizing Incremental Neural Network Plus
% ver. 1.0.1
%
% Copyright (c) 2016 Yoshihiro Nakamura
% Copyright (c) 2019 Chayut Wiwatcharakoses
% This software is released under the MIT License.
% http://opensource.org/licenses/mit-license.php

classdef SoinnPlus < handle

    properties
        % Internal variable.
        nodeDelTh;
        nodeAvgIdleDel;
        edgeAvgLtDel;
        curNodeTh;
        curEdgeTh;
		enableTracking = 1;
        minDegree = 1;
        paramEdge = 2;
        paramC = 2;
        paramAlpha = 2;
        
        % Parameter
        dimension;
        deleteNodePeriod;
        maxEdgeAge;
        
        % SOINN+ Option
        nodeFlag;
        edgeFlag;

        % Data related variable
        signalNum;
        nodes;
		trackInput;
		trackInputIdx;
        winningTimes;
        winTS;
        nodeTS;
        adjacencyMat;
        linkCreated;
        nodeDeleted;
        edgeDeleted;
        runThVariance;
        runThMean;

        % Function Handler
        deleteNoiseHandler;
        deleteEdgeHandler;
       
        
    end

    methods
        function obj = SoinnPlus(varargin)
            % constractor
            % Legacy Options: 
            % 'lambda' (default: 300)
            %       A period deleting nodes. The nodes that doesn't satisfy
            %       some condition are deleted every this period.
            % 'ageMax' (default: 50)
            %       The maximum of edges' ages. If an edge's age is more
            %       than this, the edge is deleted.
            % 'dim' (default: 2)
            %       signal's dimension
            %
            % Plus Options:
            % 'node' (default: 1)
            %       1 - Enable plus version of node deletion
            %       0 - Disable plus version of node deletion
            % 'edge' (default: 1)
            %       1 - Enable plus version of node linking
            %       0 - Disable plus version of node linking
            
            o = OptionHandler(varargin);
            obj.deleteNodePeriod = o.get('lambda', 300);
            obj.maxEdgeAge = o.get('ageMax', 50);
            obj.dimension = o.get('dim', 2);
            
            obj.nodeFlag = o.get('node', 1);  
            obj.edgeFlag = o.get('edge', 1);  
            obj.runThVariance = [0 0];
            obj.runThMean = [0 0];
            obj.linkCreated = 0;
            obj.nodes = [];
			obj.trackInput = {};
			obj.trackInputIdx = {};
            obj.winningTimes = [];
            obj.winTS = [];
            obj.nodeTS = [];
            obj.nodeDelTh = 0;
            obj.nodeAvgIdleDel = 0;
            obj.edgeAvgLtDel = 0;
            obj.adjacencyMat = sparse([]);
            obj.signalNum = 0;
            obj.nodeDeleted = 0;
            obj.edgeDeleted = 0;
 
            if obj.nodeFlag == 0
                obj.deleteNoiseHandler = @obj.deleteNoiseNodes_Original;
            else
                obj.deleteNoiseHandler = @obj.deleteNoiseNodes_Plus;
                obj.deleteNodePeriod = 1;
            end
            
            if obj.edgeFlag == 0
                obj.deleteEdgeHandler = @obj.deleteOldEdges_Original;
            else
                obj.deleteEdgeHandler = @obj.deleteOldEdges_Plus;
            end

        end

        % Main function when receiving new cases goes here ---------------
        function inputSignal(obj, signal, varargin)
            % @param {row vector} signal - new input signal
            signal = obj.checkSignal(signal);
            obj.signalNum = obj.signalNum + 1;

            % If in initialization state add node unconditionally
            if size(obj.nodes, 1) < 3 
                obj.addNode(signal, 0);
                return;
            end

            % Find the winners and calculate similarity threshold
            [winner, dists] = obj.findNearestNodes(2,signal);
            simThresholds = obj.calculateSimiralityThresholds(winner);

            % Check if the network should create the link between both
            % winner or not
            if any(obj.runThVariance == 0) 
                eFlag = 1; % Create link unconditionally if there is no edge in the network
            else
                % Check if edge between both winner should be created or
                % not
                th = obj.paramC*sqrt(obj.runThVariance/obj.linkCreated);
                noises = sum(obj.adjacencyMat > 0) < 1; 
                data = ~full(noises);
                trustLv = obj.winningTimes(winner)./max(obj.winningTimes(data)); % Calculate trustworthiness
                eFlag = any(sqrt(simThresholds).*(1-trustLv') < (obj.runThMean + th)');
            end
            
            % Add node if either one of distance greater than corresponding similarity threshold
            if any(dists > simThresholds) 
                obj.addNode(signal, 0);
            else
                
                % Add edge if the the condition is true
                if eFlag
                    isNew = obj.addEdge(winner);
                    if isNew
                        obj.linkCreated = obj.linkCreated+1;
                        preMean = obj.runThMean;
                        % Update the mean and variance of similarity
                        % threhsold
                        obj.runThMean = obj.runThMean + (sqrt(simThresholds') - obj.runThMean)./obj.linkCreated;
                        obj.runThVariance = obj.runThVariance + (sqrt(simThresholds') - preMean).*(sqrt(simThresholds') - obj.runThMean);
                    end
                end
                
                % After process updating
                obj.incrementEdgeAges(winner(1));
                winner(1) = obj.deleteEdgeHandler(winner(1));
                obj.updateWinner(winner(1), signal);
                obj.updateAdjacentNodes(winner(1), signal);
            
			end

            % Check if any node can be deleted
            if mod(obj.signalNum, obj.deleteNodePeriod) == 0
                obj.deleteNoiseHandler();
            end
        end

        function show(obj, varargin)
            % Display SOINN's network in 2D.
            % Options:
            % 'winningTimes' a flag to show winning time of each node
            % 'dim' which dimensions to show (default: [1, 2])
            % 'data' data showed with KDESOINN status (default: [])
            % 'cluster' cluster labels which is a row vector outputed by clustring()
            o = OptionHandler(varargin);
            dims = o.get('dim', [1 2]);
            winningTimeFlag = o.exist('winningTimes');
            data = o.get('data', []);
            hold on;
            %show data
            if size(data, 1) > 0
                plot(data(:,dims(1)), data(:,dims(2)), 'xb');
            end
            % show edges
            for j = 1:size(obj.adjacencyMat,2)
                for k = j:size(obj.adjacencyMat, 1)
                    if obj.adjacencyMat(k,j) > 0
                        nk = obj.nodes(k,:);
                        nj = obj.nodes(j,:);
                        plot([nk(dims(1)), nj(dims(1))], [nk(dims(2)), nj(dims(2))], 'k');
                    end
                end
            end
            % show nodes
            clusterLabels = o.get('cluster', []);
            if ~isempty(clusterLabels)
                clusterNum = max(clusterLabels);
                colors = obj.getRgbVectors(clusterNum+1);
                idx = clusterLabels == -1;
                plot(obj.nodes(idx, dims(1)), obj.nodes(idx, dims(2)), '.', 'Markersize', 20, 'MarkerEdgeColor', colors(1, :));
                for i = 1:clusterNum
                    idx = clusterLabels == i;
                    plot(obj.nodes(idx, dims(1)), obj.nodes(idx, dims(2)), '.', 'Markersize', 20, 'MarkerEdgeColor', colors(i+1, :));
                    %text(obj.nodes(idx, dims(1)), obj.nodes(idx, dims(2))-5, num2str(obj.winningTimes(idx))); %'\leftarrow sin(\pi)')
                end
            else
                plot(obj.nodes(:,dims(1)), obj.nodes(:,dims(2)), '.b','Markersize',20);
               % plot(obj.delNodes(:,dims(1)), obj.delNodes(:,dims(2)), '.r','Markersize',15);
            end
            
            % plot Winning Time
            %for i = 1:size(obj.nodes,1)
            %    text(obj.nodes(i, dims(1))+1, obj.nodes(i, dims(2)), num2str(obj.winningTimes(i))); %'\leftarrow sin(\pi)')
            %end
            
            % show winningTimes of nodes
            if winningTimeFlag
                delta = (max(max(obj.nodes(:,dims))) - min(min(obj.nodes(:,dims)))) * 0.005;
                for i = 1:size(obj.nodes,1)
                    text(obj.nodes(i,dims(1)) + delta, obj.nodes(i,dims(2)) + delta, num2str(obj.winningTimes(i)), 'BackgroundColor', [1,1,1]);
                end
            end
%             title(strcat('nodes:', num2str(size(obj.nodes,1)), ' edges:', num2str(sum(sum((obj.adjacencyMat > 0)))), ' WinTime:', num2str(sum(obj.winningTimes)) ));
           
            %title(strcat('nodes:', num2str(size(obj.nodes,1)), ' edges:', num2str(sum(sum((obj.adjacencyMat > 0)))), ' WinTime:', num2str(sum(obj.winningTimes)) , ' DN:', num2str(obj.nodeDeleted), ' DE:', num2str(obj.edgeDeleted)));
            set(gca,'XGrid','on','YGrid','on');
            hold off
            if o.get('save', false)
                saveas(gcf, o.get('savePath', ['tmp.png']))
            end
        end

        function err = topoErr(obj, Data)
            numErr = 0;
            for i = 1:size(Data, 1)
                nearIdx = obj.findNearestNodes(2, Data(i,:));
                if obj.adjacencyMat(nearIdx(1), nearIdx(2))
                    numErr = numErr + 1;
                end
            end
            err = numErr/size(Data,1);
        end
    end
%%
    methods(Hidden=true)
        function rgbs = getRgbVectors(obj, num)
            rgbs = [];
            maxBase = ceil(power(num, 1/3));
            for base = 2:maxBase
                candi = zeros(base^3, 3);
                for i = 0:base^3-1
                    candi(i+1,:) = obj.str2numArry(dec2base(i,base,3));
                end
                candi = candi / (base-1);
                for i = 1:size(candi, 1)
                    tf = true;
                    for k = 1:size(rgbs, 1)
                        if all(candi(i,:) == rgbs(k,:))
                            tf = false;
                            break;
                        end
                    end
                    if tf
                        rgbs = cat(1, rgbs, candi(i, :));
                    end
                end
            end
            rgbs = rgbs(1:num, :);
        end

        function arry = str2numArry(~, str)
            n = length(str);
            arry = zeros(1, n);
            for i = 1:n
                arry(i) = str2num(str(i));
            end
        end

        function [labels, num] = labelWithBreadthFirstSearch(obj, labels, queue, currentClusterLabel)
            num = 0;
            while ~isempty(queue)
                idx = queue(1);
                queue(1) = [];
                if labels(idx) == 0
                    num = num + 1;
                    labels(idx) = currentClusterLabel;
                    queue = cat(2, queue, find(obj.adjacencyMat(idx, :) > 0));
                end
            end
        end

        function signal = checkSignal(obj, signal)
            s = size(signal);
            if s(1) == 1
                if s(2) == obj.dimension
                    return;
                else
                    error('Soinn:checkSignal:dimError', 'The dimension of input signal is not valid.');
                end
            else
                if s(2) == 1
                    signal = obj.checkSignal(signal');
                else
                    error('Soinn:checkSignal:notVector', 'Input signal have to be a vector.');
                end
            end
        end

        function addNode(obj, signal, varargin)
            num = size(obj.nodes, 1);
            obj.nodes(num+1,:) = signal;
            obj.winningTimes(num+1) = 1;
            obj.winTS(num+1) = obj.signalNum;
            obj.nodeTS(num+1) = obj.signalNum;

            if num == 0
                obj.adjacencyMat(1,1) = 0;
            else
                obj.adjacencyMat(num+1,:) = zeros(1, num);
                obj.adjacencyMat(:,num+1) = zeros(num+1, 1);
            end
			
			if obj.enableTracking
				obj.trackInput{num+1} = signal;
				obj.trackInputIdx{num+1} = obj.signalNum;
			end
        end

        function [indexes, sqDists] = findNearestNodes(obj, num, signal)
            indexes = zeros(num, 1);
            sqDists = zeros(num, 1);
            %D = sum(((obj.nodes - repmat(signal, size(obj.nodes, 1), 1)).^2), 2);
            D = sum((bsxfun(@minus,obj.nodes,signal).^2), 2);
            for i = 1:num
                [sqDists(i), indexes(i)] = min(D);
                D(indexes(i)) = inf;
            end
        end

        function simThresholds = calculateSimiralityThresholds(obj, nodeIndexes)
            simThresholds = zeros(length(nodeIndexes), 1);
            for i = 1: length(nodeIndexes)
                simThresholds(i) = obj.calculateSimiralityThreshold(nodeIndexes(i));
            end
        end

        function threshold = calculateSimiralityThreshold(obj, nodeIndex)
            if any(obj.adjacencyMat(:,nodeIndex))
                pals = obj.nodes(obj.adjacencyMat(:,nodeIndex) > 0,:);
                D = sum(((pals - repmat(obj.nodes(nodeIndex,:), size(pals, 1), 1)).^2), 2);
                threshold = max(D);
            else
                [~, sqDists] = obj.findNearestNodes(2, obj.nodes(nodeIndex, :));
                threshold = sqDists(2);
                
            end
        end

        function isNew = addEdge(obj, nodeIndexes)
            if obj.adjacencyMat(nodeIndexes(1), nodeIndexes(2)) || obj.adjacencyMat(nodeIndexes(2), nodeIndexes(1))
                isNew = true;
            else
                isNew = false;
            end
            obj.adjacencyMat(nodeIndexes(1), nodeIndexes(2)) = 1;
            obj.adjacencyMat(nodeIndexes(2), nodeIndexes(1)) = 1;
            
        end

        function updateWinner(obj, winnerIndex, signal)
            % @param {int} winnerIndex - hte index of winner
            % @param {row vector} signal - inputted new signal
            obj.winningTimes(winnerIndex) = obj.winningTimes(winnerIndex) + 1;
            w = obj.nodes(winnerIndex,:);
            obj.nodes(winnerIndex, :) = w + (signal - w)./obj.winningTimes(winnerIndex);
            obj.winTS(winnerIndex) = obj.signalNum;
			
			if obj.enableTracking
				obj.trackInput{winnerIndex} = [obj.trackInput{winnerIndex}; signal];
				obj.trackInputIdx{winnerIndex} = [obj.trackInputIdx{winnerIndex}; obj.signalNum];
			end
        end

        function updateAdjacentNodes(obj, winnerIndex, signal)
            pals = find(obj.adjacencyMat(:,winnerIndex) > 0);
            for i = 1:length(pals)
                w = obj.nodes(pals(i),:);
                obj.nodes(pals(i), :) = w + (signal - w)./(100 * obj.winningTimes(pals(i)));
            end
        end

        function incrementEdgeAges(obj, winnerIndex)
            indexes = find(obj.adjacencyMat(:,winnerIndex) > 0);
            for i = 1: length(indexes)
                obj.incrementEdgeAge(winnerIndex, indexes(i));
            end
        end

        function incrementEdgeAge(obj, i, j)
            obj.adjacencyMat(i, j) = obj.adjacencyMat(i, j) + 1;
            obj.adjacencyMat(j, i) = obj.adjacencyMat(j, i) + 1;
        end

        function setEdgeAge(obj, i, j, value)
            obj.adjacencyMat(i, j) = value;
            obj.adjacencyMat(j, i) = value;
        end
        
        function winnerIndex = deleteOldEdges_Original(obj, winnerIndex)
            indexes = find(obj.adjacencyMat(:,winnerIndex) > obj.maxEdgeAge + 1); % 1 expresses that there is an edge.
            deletedNodeIndexes = [];
            for i = 1: length(indexes)
                obj.edgeDeleted = obj.edgeDeleted+1;
                obj.setEdgeAge(indexes(i), winnerIndex, 0);
                obj.setBond(indexes(i), winnerIndex, 0);
                if ~any(obj.adjacencyMat(:,indexes(i)))
                    deletedNodeIndexes = cat(1, deletedNodeIndexes, indexes(i));
                end
            end
            winnerIndex = winnerIndex - sum(deletedNodeIndexes < winnerIndex);
            obj.deleteNodes(deletedNodeIndexes);
        end

        function edgeAge = collectClusterEdgeAge(obj, seed)
            G = graph(obj.adjacencyMat);
          
            T = dfsearch(G, seed,'allevents');

            eList = unique(T.EdgeIndex);

            eList(isnan(eList)) = [];
            edgeAge = G.Edges.Weight(eList);

        end

        function winnerIndex = deleteOldEdges_Plus(obj, winnerIndex)
            % Collect all edges in cluster
            edgeAge = obj.collectClusterEdgeAge(winnerIndex);

            edge = obj.adjacencyMat(:,winnerIndex);
 
            c = prctile(edgeAge,75);
            th = obj.paramEdge*iqr(edgeAge);
            
            if th == 0
                return;
            end
            
            curTh = c + th;
            ratio = obj.edgeDeleted / ( obj.edgeDeleted + length(edgeAge));

            % Check if there are any edges to be deleted
            delThreshold = (obj.edgeAvgLtDel*ratio + curTh*(1-ratio) );
            delFlag = edge > delThreshold;
            indexes = find(delFlag);
            obj.curNodeTh = delThreshold;
            
            % Update average lifetime of deleted edges
            if ~isempty(indexes)
                obj.edgeAvgLtDel = (obj.edgeDeleted*obj.edgeAvgLtDel + sum(edge(delFlag))) / (obj.edgeDeleted + length(indexes));
            end
            
            obj.edgeDeleted = obj.edgeDeleted + length(indexes);
            deletedNodeIndexes = [];
            for i = 1: length(indexes)
                
                obj.setEdgeAge(indexes(i), winnerIndex, 0);
                if ~any(obj.adjacencyMat(:,indexes(i)))
                    deletedNodeIndexes = cat(1, deletedNodeIndexes, indexes(i));
                end
            end
            
            % Update winner index according to the deleted node
            winnerIndex = winnerIndex - sum(deletedNodeIndexes < winnerIndex);
            obj.deleteNodes(deletedNodeIndexes);
        end

        
   
        function deleteNodes(obj, indexes)
            % record deleted node's signatures and remove signature
            % recoders.

            % remove the nodes.
            obj.nodes(indexes,:) = [];
            obj.winningTimes(indexes) = [];
            obj.adjacencyMat(indexes, :) = [];
            obj.adjacencyMat(:, indexes) = [];
            obj.winTS(indexes) = [];
            obj.nodeTS(indexes) = [];

			if obj.enableTracking
				obj.trackInput(indexes) = [];
				obj.trackInputIdx(indexes) = [];
			end

            obj.nodeDeleted = obj.nodeDeleted + sum(indexes);

        end

        function deleteNoiseNodes_Original(obj, varargin)
            noises = sum(obj.adjacencyMat > 0) < obj.minDegree;
            
            obj.deleteNodes(noises);
        end
        
        function deleteNoiseNodes_Plus(obj, varargin)
            noises = sum(obj.adjacencyMat > 0) < obj.minDegree; 
            data = ~full(noises);
 
            IT = obj.signalNum - obj.winTS;
            UT = IT./obj.winningTimes; % unutility

            [TF, l,u,c] = isoutlier(UT(data));
            th = obj.paramAlpha*((u - c)/3);
            curTh = c + th;

            % Check if there are any nodes should be deleted
            ratio = obj.nodeDeleted / ( obj.nodeDeleted + sum(data));
            noiseLv = sum(noises) / size(obj.nodes, 1);
            delThreshold = (obj.nodeDelTh*(ratio) + curTh*(1-ratio)*(1-noiseLv) );
            inactiveIdx = UT > delThreshold;
            obj.curNodeTh = delThreshold; % tracking the main deleted threshold

            if any(inactiveIdx & noises)
                % Tracking the average deleted idle time and unutility 
                obj.nodeAvgIdleDel = (obj.nodeDeleted*obj.nodeAvgIdleDel + sum(IT(inactiveIdx & noises))) / (obj.nodeDeleted + sum(inactiveIdx & noises));
                obj.nodeDelTh = (obj.nodeDeleted*obj.nodeDelTh + sum(UT(inactiveIdx & noises))) / (obj.nodeDeleted + sum(inactiveIdx & noises));
            end
            obj.deleteNodes(inactiveIdx & noises);
        end
    end
end

File: soinn/soinnplus.py
from math import sqrt
from collections import deque
import numpy as np
from soinn.soinn import Soinn
from scipy import stats
from scipy.sparse import dok_matrix, tril
from scipy.sparse.csgraph import breadth_first_order

class NodePlus:
    def __init__(self, signal, timestamp):
        self.original_weight = signal
        self.winning_ts = timestamp # Last winning timestamp
        self.node_ts = timestamp # Created timestamp



class SoinnPlus(Soinn):
    def __init__(self, delete_node_period=300, max_edge_age=50,
                 init_node_num=3):
        
        super().__init__(delete_node_period, max_edge_age, init_node_num)

        self.nodeDelTh = 0
        self.nodeAvgIdleDel = 0
        self.edgeAvgLtDel = 0
        self.curNodeTh = None
        self.curEdgeTh = None
        self.enableTracking = 1
        self.minDegree = 1
        self.paramEdge = 2
        self.paramC = 2
        self.paramAlpha = 2

        self.runThVariance = np.array([0, 0], dtype=np.float64)
        self.runThMean = np.array([0, 0], dtype=np.float64)
        self.linkCreated = 0
        self.nodeDeleted = 0
        self.edgeDeleted = 0

        # self.nodes = []
        self.winning_times = []
        self.winning_ts = []
        self.node_ts = []

    def __check_signal(self, signal: np.ndarray):
        """
        check type and dimensionality of an input signal.
        If signal is the first input signal, set the dimension of it as
        self.dim. So, this method have to be called before calling functions
        that use self.dim.
        :param signal: an input signal
        """
        if isinstance(signal, list):
            signal = np.array(signal)
        if not(isinstance(signal, np.ndarray)):
            raise TypeError()
        if len(signal.shape) != 1:
            raise TypeError()
        if self.dim is None:
            self.dim = signal.shape[0]
        else:
            if signal.shape[0] != self.dim:
                raise TypeError()
        return signal

    def __add_node(self, signal: np.ndarray):
        n = self.nodes.shape[0]
        self.nodes.resize((n + 1, self.dim), refcheck=False)
        self.nodes[-1, :] = signal
        self.winning_times.append(1)
        self.adjacent_mat.resize((n + 1, n + 1))

        ## Soinn Plus
        self.winning_ts.append(self.num_signal)
        self.node_ts.append(self.num_signal)
    
    def __find_nearest_nodes(self, num: int, signal: np.ndarray):
        n = self.nodes.shape[0]
        indexes = [0] * num
        sq_dists = [0.0] * num
        D = np.sum((self.nodes - np.array([signal] * n))**2, 1)
        for i in range(num):
            indexes[i] = np.nanargmin(D)
            sq_dists[i] = D[indexes[i]]
            D[indexes[i]] = float('nan')
        return indexes, sq_dists

    def __calculate_similarity_thresholds(self, node_indexes):
        sim_thresholds = []
        for i in node_indexes:
            pals = self.adjacent_mat[i, :]
            if len(pals) == 0:
                idx, sq_dists = self.__find_nearest_nodes(2, self.nodes[i, :])
                sim_thresholds.append(sq_dists[1])
            else:
                pal_indexes = []
                for k in pals.keys():
                    pal_indexes.append(k[1])
                sq_dists = np.sum((self.nodes[pal_indexes] - np.array([self.nodes[i] * len(pal_indexes)]))**2, 1)
                sim_thresholds.append(np.max(sq_dists))
        return sim_thresholds

    def __add_edge(self, node_indexes):
        return self.__set_edge_weight(node_indexes, 1)

    def __increment_edge_ages(self, winner_index):
        for k, v in self.adjacent_mat[winner_index, :].items():
            self.__set_edge_weight((winner_index, k[1]), v + 1)

    def __collect_cluster_edge_age(self, seed):
        queue = deque()
        queue.append(seed)
        edge_list = []
        seen_set = set()

        search_mat = self.adjacent_mat.copy()

        while queue:
            node = queue.popleft()
            if node in seen_set:
                continue

            seen_set.add(node)
            for k, w in self.adjacent_mat[node, :].items():
                if w > 0 and search_mat[node, k[1]]:
                    edge_list.append(w)
                    queue.append(k[1])
                    search_mat[node, k[1]] = 0
                    search_mat[k[1], node] = 0

        return edge_list

    def __delete_old_edges(self, winner_index):
        edge_ages = self.__collect_cluster_edge_age(winner_index)
        q3, q1 = np.percentile(edge_ages, [75, 25])
        th = self.paramEdge*(q3-q1)

        if th == 0:
            return winner_index
        
        curTh = q3 + th
        ratio = self.edgeDeleted / (self.edgeDeleted + len(edge_ages))

        delThreshold = (self.edgeAvgLtDel*ratio) + curTh*(1-ratio)

        candidates = []
        sum_del_edges = 0
        for k, v in self.adjacent_mat[winner_index, :].items():
            if v > delThreshold:
                candidates.append(k[1])
                sum_del_edges += v
                self.__set_edge_weight((winner_index, k[1]), 0)
        
        if candidates:
            self.edgeAvgLtDel = (self.edgeDeleted*self.edgeAvgLtDel + sum_del_edges) / (self.edgeDeleted + len(candidates))
        
        self.edgeDeleted += len(candidates)
        
        delete_indexes = []
        for i in candidates:
            if len(self.adjacent_mat[i, :]) == 0:
                delete_indexes.append(i)

        self.__delete_nodes(delete_indexes)
        delete_count = sum([1 if i < winner_index else 0 for i in delete_indexes])
        return winner_index - delete_count

    def __set_edge_weight(self, index, weight):
        is_new_edge = False
        if self.adjacent_mat[index[0], index[1]] == 0 or self.adjacent_mat[index[1], index[0]] == 0:
            is_new_edge = True
        self.adjacent_mat[index[0], index[1]] = weight
        self.adjacent_mat[index[1], index[0]] = weight

        return is_new_edge

    def __update_winner(self, winner_index, signal):
        self.winning_times[winner_index] += 1
        w = self.nodes[winner_index]
        self.nodes[winner_index] = w + (signal - w)/self.winning_times[winner_index]
        self.winning_ts[winner_index] = self.num_signal

    def __update_adjacent_nodes(self, winner_index, signal):
        pals = self.adjacent_mat[winner_index]
        for k in pals.keys():
            i = k[1]
            w = self.nodes[i]
            self.nodes[i] = w + (signal - w)/(100 * self.winning_times[i])

    def __delete_nodes(self, indexes):
        if not indexes:
            return
        n = len(self.winning_times)
        self.nodes = np.delete(self.nodes, indexes, 0)
        remained_indexes = list(set([i for i in range(n)]) - set(indexes))
        self.winning_times = [self.winning_times[i] for i in remained_indexes]

        ### Soinn Plus
        self.winning_ts = [self.winning_ts[i] for i in remained_indexes] 
        self.node_ts = [self.node_ts[i] for i in remained_indexes]

        self.__delete_nodes_from_adjacent_mat(indexes, n, len(remained_indexes))

    def __delete_nodes_from_adjacent_mat(self, indexes, prev_n, next_n):
        while indexes:
            next_adjacent_mat = dok_matrix((prev_n, prev_n))
            for key1, key2 in self.adjacent_mat.keys():
                if key1 == indexes[0] or key2 == indexes[0]:
                    continue
                if key1 > indexes[0]:
                    new_key1 = key1 - 1
                else:
                    new_key1 = key1
                if key2 > indexes[0]:
                    new_key2 = key2 - 1
                else:
                    new_key2 = key2
                #Because dok_matrix.__getitem__ is slow,
                #access as dictionary.
                next_adjacent_mat[new_key1, new_key2] = super(dok_matrix, self.adjacent_mat).__getitem__((key1, key2))
            self.adjacent_mat = next_adjacent_mat.copy()
            indexes = [i-1 for i in indexes]
            indexes.pop(0)
        self.adjacent_mat.resize((next_n, next_n))

    def __delete_noise_nodes(self):
        data = np.where((sum(self.adjacent_mat ) != 0).toarray())[1]
        noise = [ele for ele in range(self.adjacent_mat.shape[0]) if ele not in data]

        IT = np.subtract(self.num_signal, self.winning_ts) 
        UT = np.divide(IT, self.winning_times)
        ut_data = UT[data]
        med = np.median(ut_data)
        mad = stats.median_abs_deviation(ut_data)
        curTh = med + 2 * mad

        ### Check if any node should be deleted
        ratio = self.nodeDeleted / (self.nodeDeleted + len(data))
        noise_lv = len(noise) / len(self.nodes)
        delThreshold = self.nodeDelTh*ratio + (curTh*(1-ratio)*(1-noise_lv))
        self.curNodeTh = delThreshold



        ### Original section
        n = len(self.winning_times)
        noise_indexes = []
        for i in range(n):
            if len(self.adjacent_mat[i, :]) < self.min_degree and UT[i] > delThreshold:
                noise_indexes.append(i)
        
        if noise_indexes:
            self.nodeAvgIdleDel = (self.nodeDeleted*self.nodeAvgIdleDel + sum(IT[noise_indexes])) / (self.nodeDeleted + len(noise_indexes))
            self.nodeDelTh = (self.nodeDeleted*self.nodeDelTh + sum(UT[noise_indexes])) / (self.nodeDeleted + len(noise_indexes)) 

        self.__delete_nodes(noise_indexes)
        self.nodeDeleted += len(noise_indexes)
    
    def show_network_stats(self):
        print("Number of Nodes:", len(self.nodes))
        print("Running Variance Threshold:", self.runThVariance)
        print("Running Mean:", self.runThMean)
        print("Link Created:", self.linkCreated)
        print("Node Deleted:", self.nodeDeleted)
        print("Edge Deleted:", self.edgeDeleted)


    def fit(self, X):
        """
        train data in batch manner
        :param X: array-like or ndarray
        """
        self._reset_state()
        for x in X:
            self.input_signal(x)
        # self.labels_ = self.__label_samples(X)
        return self

    def input_signal(self, signal: np.ndarray):
        """
        Input a new signal one by one, which means training in online manner.
        fit() calls __init__() before training, which means resetting the
        state. So the function does batch training.
        :param signal: A new input signal
        :return:
        """

        signal = self.__check_signal(signal)
        self.num_signal += 1

        if self.nodes.shape[0] < self.init_node_num:
            self.__add_node(signal)
            return

        winners, dists = self.__find_nearest_nodes(2, signal)
        sim_thresholds = self.__calculate_similarity_thresholds(winners)

        print("Winners:", winners)

        run_vars = [x == 0 for x in self.runThVariance]


        # Check if the network should create the link
        if all(run_vars):
            print("Force Create Link", run_vars)
            eFlag = 1
        else:
            th = self.paramC*np.sqrt(self.runThVariance/self.linkCreated)
            # get data index
            data = np.where((sum(self.adjacent_mat ) != 0).toarray())[1]
            trustLV = np.array(self.winning_times)[winners] / np.array(self.winning_times)[data].max()
            condition = []
            for i in range(len(sim_thresholds)):
                value = sim_thresholds[i] * (1-trustLV[i])
                th1 = self.paramC*sqrt(self.runThVariance[i] / self.linkCreated)
                result = value < self.runThMean[i] + th1
                condition.append(result)
                
            eFlag = any(condition)
            print("Link threshold", th)
            print("trust LV", trustLV)
            ########## Reach here for implementation <<<<<<<<<-----------------
            # trustLv = self.winningTimes(winners) / max(self.winningTimes[data])
            # T = [L[i] for i in Idx] get all winning time in the node
            # winning_times = np.array([self.nodes[idx].winningTimes for idx in winners])
            # max_trust = max([self.nodes[idx].winningTimes for idx in data])
            # trustLv = winning_times / max_trust # max([node.winningTimes for node in self.nodes])
            # link_value = simThresholds * (1-trustLv)
            # threshold = (self.runThMean + th)
            # eFlag = any(link_value < threshold)

        if (dists[0] > sim_thresholds[0] or dists[1] > sim_thresholds[1]): # and not eFlag:
            self.__add_node(signal)
        else:
            ### New Rule 
            # if eFlag allow then add edge
            if eFlag:
                is_new = self.__add_edge(winners)
                if is_new:
                    self.linkCreated += 1
                    pre_mean = self.runThMean

                    # Update mean and var of sim threshold
                    self.runThMean = self.runThMean + (sim_thresholds - self.runThMean) / self.linkCreated
                    self.runThVariance = self.runThVariance + (sim_thresholds - pre_mean) * (sim_thresholds - self.runThMean)

            self.__increment_edge_ages(winners[0])
            winners[0] = self.__delete_old_edges(winners[0])
            self.__update_winner(winners[0], signal)
            self.__update_adjacent_nodes(winners[0], signal)

        if self.num_signal % self.delete_node_period == 0:
            self.__delete_noise_nodes()

        self.show_network_stats()